### PR TITLE
[Validator][ConstraintViolation] Implemented Serializable

### PR DESCRIPTION
**Pull request by alexmarucci**
Implemented serializable and ignore the root property to avoid exception.

| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | yes
| New feature?  | no 
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes
| Fixed tickets | #24537 
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by Simperfit** [comment]
If it's a bugfix it should be in the lower branch where the bug exists.

Could you please check why the tests are not passing ?

File: src/Symfony/Component/Form/Tests/FormErrorSerializableTest.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+use Symfony\Component\Form\FormError;
+use Symfony\Component\Validator\ConstraintViolation;
+
+class FormErrorSerializableTest extends TestCase
+{
+    public function testSerializeFormError()
+    {
+        if (!class_exists(ConstraintViolation::class)) {
+            $this->markTestSkipped('Validator component required.');
+        }
```
Comment (Simperfit) [inline src/Symfony/Component/Form/Tests/FormErrorSerializableTest.php]: LB

**Review by Simperfit — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Validator/ConstraintViolation.php [hunk 2]
```diff
@@ -201,4 +201,28 @@ public function getCode()
     {
         return $this->code;
     }
+
+    public function serialize()
+    {
+        return serialize(array(
+                $this->message,
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/ConstraintViolation.php]: indentation is one level too much

File: src/Symfony/Component/Form/Tests/FormErrorSerializableTest.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            array()
+        );
+
+        $form = $formBuilder->getForm();
+        $simpleXMLElement = new \SimpleXMLElement('<foo></foo>');
+        $cause = new ConstraintViolation('Error 1!', null, array(), $simpleXMLElement, '', null, null, '');
+        $form->addError(new FormError('Error 1!', null, array(), null, $cause));
+        $formErrors = $form->getErrors();
+        $specificFormError = $formErrors->current();
+        dump($cause->serialize());
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Tests/FormErrorSerializableTest.php]: ?

Comment (alexmarucci) [reply src/Symfony/Component/Form/Tests/FormErrorSerializableTest.php]: Sure, I can use the FormError directly without instantiate a form object.

line 32: try to call the serialize function as on Travis ( see below ) for some reason the ConstraintViolation class has been overrided.

https://travis-ci.org/symfony/symfony/jobs/313999394

**Comment by nicolas-grekas** [comment]
Reading the linked issue, I understand why this has been submitted  (`$this->root` may be unserializable), but from a local pov, not sure it makes sense.
Any other idea?

**Comment by xabbuh** [comment]
moving to the 3.4 milestone as the last bugfix release for 3.3 was published today

**Comment by alexmarucci** [comment]
Another solution could be to check in FormError if the cause itself is serializable, as @ogizanagi suggest in the linked issue.
Something like:
`$this->isSerializable($this->cause) ? $this->cause : null,`
That would prevent the Exception to be trown. 
isSerializable would be a private function inside FormError class like so;
`private function isSerializable($property)
    {
        try {   
            serialize( $property );
            return true;
        } catch (\Exception $e) {
            return false;
        }
    }`

What do you guys think about that solution ?

**Comment by ogizanagi** [comment]
Unfortunately, I don't have much more appart making `ConstraintViolation` implement `\Serializable` as here and ignore `FormError::$cause`, `ConstraintViolation::$cause,$root` in the serialization, unless scalar or implementing `\Serializable`. 
Or implement `Form::__sleep()` or `FormConfigInterface::__sleep()`. 
But that's more workarounds than proper fixes :/

**Comment by alexmarucci** [comment]
Hey Guys,

Couldn't find the reason why test are failing on travis. If someone has any idea is more than welcome.

**Comment by ro0NL** [comment]
Note you can already serialize the violation list using the `Serializer`, e.g. https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Serializer/Normalizer/ConstraintViolationListNormalizer.php

Which will work with any `ConstraintViolationListInterface` object, not just `ConstraintViolation`. But implementing serializable here doesnt hurt i guess :)

**Comment by fabpot** [comment]
Do we still need this PR after @ro0NL's comment?
If yes, this should be rebased on master as this is a new feature.

**Comment by ogizanagi** [comment]
I missed @ro0NL 's comment on this PR. I might be misinterpreting, but I fail to understand how it relates to the original issue. Yes, constraint violations can be serialized using the Symfony serializer.
But the original issue is about serializing a `FormError` using PHP as `FormError` implements `\Serializable`.

However, I don't know the use-case, and perhaps what's wrong here is letting users think a `FormError` can be safely serialized using PHP (not sure why the interface was added on this class in the first place? Probably for older version of the profiler). Should we deprecate the `\Serializable` implementation in `FormError`?

**Comment by xabbuh** [comment]
Aren't form errors serialised when the profile is written to the filesystem?

**Comment by ogizanagi** [comment]
@xabbuh : I don't think that's true anymore. We serialize `Data` (VarDumper) clones.

**Comment by nicolas-grekas** [comment]
I'm closing here as serializing is not a primary use case of the class. Using VarDumper for cloning is used internally - and should/could be used in similar userland cases IMHO - (using the Serializer component would work also).

Thanks @alexmarucci for helping this move forward.

**Pull request closed without merging**